src/distributions/microsoft/installer.ts

```diff
@@ -0,0 +1,122 @@
+import { JavaDownloadRelease, JavaInstallerOptions, JavaInstallerResults } from '../base-models';
+import semver from 'semver';
+import { extractJdkFile, getDownloadArchiveExtension, isVersionSatisfies } from '../../util';
+import * as core from '@actions/core';
+import { MicrosoftVersion, PlatformOptions } from './models';
+import * as tc from '@actions/tool-cache';
+import fs from 'fs';
+import path from 'path';
+
+const supportedPlatform = `'linux', 'macos', 'windows'`;
+
+const supportedArchitecture = `'x64', 'armv7', 'aarch64'`;
+
+export class MicrosoftDistributions extends JavaBase {
+  constructor(installerOptions: JavaInstallerOptions) {
+    super('Microsoft', installerOptions);
+  }
+
+  protected async downloadTool(javaRelease: JavaDownloadRelease): Promise<JavaInstallerResults> {
+    core.info(
+      `Downloading Java ${javaRelease.version} (${this.distribution}) from ${javaRelease.url} ...`
+    );
+    const javaArchivePath = await tc.downloadTool(javaRelease.url);
+
+    core.info(`Extracting Java archive...`);
+    const extension = getDownloadArchiveExtension();
+    const extractedJavaPath = await extractJdkFile(javaArchivePath, extension);
+
+    const archiveName = fs.readdirSync(extractedJavaPath)[0];
+    const archivePath = path.join(extractedJavaPath, archiveName);
+
+    const javaPath = await tc.cacheDir(
+      archivePath,
+      this.toolcacheFolderName,
+      this.getToolcacheVersionName(javaRelease.version),
+      this.architecture
+    );
+
+    return { version: javaRelease.version, path: javaPath };
+  }
+
+  protected async findPackageForDownload(range: string): Promise<JavaDownloadRelease> {
+    if (this.architecture !== 'x64' && this.architecture != 'aarch64') {
+      throw new Error(`Unsupported architecture: ${this.architecture}`);
+    }
+    const availableVersionsRaw = await this.getAvailableVersions();
+
+    const opts = this.getPlatformOption();
+    const availableVersions = availableVersionsRaw.map(item => ({
+      url: `https://aka.ms/download-jdk/microsoft-jdk-${item.fullVersion}-${opts.os}-${this.architecture}.${opts.archive}`,
+      version: this.convertVersionToSemver(item)
+    }));
+
+    const satisfiedVersion = availableVersions
+      .filter(item => isVersionSatisfies(range, item.version))
+      .sort((a, b) => -semver.compareBuild(a.version, b.version))[0];
+
+    if (!satisfiedVersion) {
+      const availableOptions = availableVersions.map(item => item.version).join(', ');
+      const availableOptionsMessage = availableOptions
+        ? `\nAvailable versions: ${availableOptions}`
+        : '';
+      throw new Error(
+        `Could not find satisfied version for semver ${range}. ${availableOptionsMessage}`
+      );
+    }
+
+    return satisfiedVersion;
+  }
+
+  private async getAvailableVersions(): Promise<MicrosoftVersion[]> {
+    console.time('microsoft-retrieve-available-versions');
+
+    // TODO get these dynamically!
```

Comment:
Do we know what it would take to get the versions dynamically?

Reply:
Unfortunately there is no easy way today other than scraping HTML which is very fragile (and annoying to implement)

I'd rather have a fixed setting for now.

Reply:
Yep, fixed setting is fine to get started.  Was just curious if we had a lead on next steps

Reply:
We plan to publish an API, but it is unclear we will get that done before the January PSU. A hardcoded version list, for now, is fine. Besides, Java updates happen 4 times a year, once every quarter. It is not that bad :-)

Reply:
Ok awesome, having an API will be good.  The main functionality this unblocks is being able to install any version that's been published from the whole history.  Regarding updates, people expect to be able to use the newest versions of tools with Actions as soon as they come out, so that also makes sure we're always up-to-date.

Not sure what your plans are for an API, but other tools (such as Node) just host a JSON file with version info and links to the downloads.  [Example](https://nodejs.org/dist/index.json)